### ISSUE TITLE
Error handling from forcing a Segment

### ISSUE THREAD

**Issue by AdamChlupacek** [functional-streams-for-scala/fs2#1128]
To carry on from the talk in the PR #1124 . 

The fact that we try to recover from errors within a `Segment.map` is doable for  single streams. But once we introduce any kind of concurrency (applies for deterministic merging as well) it all falls apart. The issue as I outlined before is that when the segment gets to the end of the stream we do not know from where it originated, as such in case of an exception while forcing a segment, we have no idea what errorHandler to call.

Lets assume this code: 

```scala
val s1: Stream[F, Int] = ???
val s2: Stream[F, Int] = ???
val s3: Stream[F, Int] = ???

Stream(
 s1.map(_ => throw new Throwable("Boom")).handleErrorWith(_ => Stream.empty), 
 s2.handleErrorWith(_ => Stream.emit(1)),
 s3
).joinUnbounded
```

Now we force the segments at the end of the stream. Where we already have no notion of which segment came from where. Which of error handler should be called to handle this issue? 

I can imagine we can figure out some tagging style that would allow us to distinguish between different elements for combinators defined inside of fs2, but once we allow users passing `Segment` around, we cannot track from where the segment came from.

As such I see we have two options: 

1) Once we encounter an error from forcing a segment, we close all scopes and terminate. 

2) We use the error handler that is closest to the point where we can track the `Segment`. As to demonstrate on a example: 

```scala
Stream(???)
.joinUnbounded
.handleErrorWith(err => Stream.emit(1)) // <- This handler would be called
```

The number 2 could lead to really hard to debug behaviour, as in some cases would the error handlers from the inner stream take effect and sometimes the error handler from outside of join would take effect.

**Comment by AdamChlupacek** [comment]
I gave this a bit more thought and I think that number one is the only valid solution. It is the only one that will behave the same in all situations and thus the `Stream` will remain compossible and easy to reason about.

if we were to go with number 2 the behaviour of which handler would be call will depend on the whole structure of the `Stream`. Ie if you have multiple `.join`s where after each one you have the `handlerErrorWith` we would call only the last `handlerErrorWith` which would be completely different behaviour to when the Streams would be joined only once and then run separately.

Now another issue is that sometimes we force the `Segment` a bit earlier than the end of the program. ie `flatMap`, `unconsChunk` and such. This once again will make a mess out of the order of error handlers. Currently resources are correctly cleared in case of any kind of `uncons` so that part is fine.

Not really sure what to do here as this means that any `Bind` of `FreeC` would have to stop the whole stream in case an error happened while evaluating the `f`. Maybe we could leave it as it is, but add somewhere a huge disclaimer that if people do throw inside pure code in some cases their error handling may be skipped?

**Comment by mpilquist** [comment]
Some miscellaneous thoughts:
# We should make it clear that throwing from pure code is discouraged (our README example that's existed since early scalaz-stream releases throws from a call to `map`, so we can do better here).
# `Segment` naturally makes error handling hard to reason about as it makes exceptions non-local.
# If a `Segment` throws when forced, it shouldn't impact invariants around scopes and resources -- e.g., scopes should be released in LIFO order no matter what our solution is. The status quo violates this stipulation.

**Comment by AdamChlupacek** [comment]
Okay. I will get together a PR where I will sort out the order of resources closing. But I will leave the `handleErrorWith` behaviour as it is. ie the error handler to the place where the segment was forced.

**Comment by pchlupacek** [comment]
I think it is completely ok if we follow laziness of Segment and just raise the exception where it is finally evaluated and not where user sort of `expects` it to be. 
Only thing we have to guarantee is that exception is raised somewhere and finalisers are run in correct order.

**Comment by pchlupacek** [comment]
hence the segment is now out of interpreter this is no longer relevant right ?

**Comment by AdamChlupacek** [comment]
Yes, this is no longer relevant, closing.
